pkg/templates/nstemplatetiers/nstemplatetier_generator_whitebox_test.go

```diff
@@ -0,0 +1,114 @@
+package nstemplatetiers
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/codeready-toolchain/host-operator/pkg/apis"
+	testnstemplatetiers "github.com/codeready-toolchain/host-operator/test/templates/nstemplatetiers"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseAllRevisions(t *testing.T) {
+
+	t.Run("ok", func(t *testing.T) {
+		// using the `metadata.yaml` test asset
+		metadata, err := testnstemplatetiers.Asset("metadata.yaml")
+		require.NoError(t, err)
+		// when
+		revisions, err := parseAllRevisions(metadata)
+		// then
+		require.NoError(t, err)
+		require.Len(t, revisions, 2)
+		require.NotContains(t, "foo", revisions) // make sure that the `foo: bar` entry was ignored
+		assert.Equal(t, "123456a", revisions["advanced"]["code"])
+		assert.Equal(t, "123456b", revisions["advanced"]["dev"])
+		assert.Equal(t, "123456c", revisions["advanced"]["stage"])
+		assert.Equal(t, "123456d", revisions["basic"]["code"])
+		assert.Equal(t, "123456e", revisions["basic"]["dev"])
+		assert.Equal(t, "123456f", revisions["basic"]["stage"])
+	})
+
+	t.Run("failures", func(t *testing.T) {
+
+		t.Run("unparseable content", func(t *testing.T) {
+			// given
+			asset := []byte("foo::bar")
+			// when initializing the generator with the production Asset
+			_, err := parseAllRevisions(asset)
+			// then
+			require.Error(t, err)
+			assert.Contains(t, err.Error(), "unable to parse all template revisions: yaml: unmarshal errors")
+		})
+
+		t.Run("invalid key format", func(t *testing.T) {
+			// given
+			asset := []byte("foo: bar")
+			// when initializing the generator with the production Asset
+			_, err := parseAllRevisions(asset)
+			// then
+			require.Error(t, err)
+			assert.Contains(t, err.Error(), "invalid namespace template filename. Expected format: '<tier_kind>-<namespace_kind>', got foo")
+		})
+
+		t.Run("invalid value format", func(t *testing.T) {
+			// given
+			asset := []byte("foo-bar: true")
+			// when initializing the generator with the production Asset
+			_, err := parseAllRevisions(asset)
+			// then
+			require.Error(t, err)
+			assert.Contains(t, err.Error(), "invalid namespace template filename revision for 'foo-bar'. Expected a string, got a bool ('true')")
+		})
+
+	})
+
+}
+
+func TestNewNSTemplateTierGenerator(t *testing.T) {
+	s := scheme.Scheme
+	err := apis.AddToScheme(s)
+	require.NoError(t, err)
+
+	t.Run("ok", func(t *testing.T) {
+		// when initializing the generator with the production Asset
+		g, err := NewNSTemplateTierGenerator(s, Asset)
+		// then
+		require.NoError(t, err)
+		// verify that there are tiers with revisions (whatever their values)
+		require.NotEmpty(t, g.revisions)
+		for tier := range g.revisions {
+			assert.NotEmpty(t, g.revisions[tier])
+		}
+	})
+
```

Comment:
may we have also a test that calls `g.NewNSTemplateTiers(namespace)` for production files? For now, there is no way we could verify that all prod yaml files have correct formatting so it won't fail while deploying to cluster, right? I know that it should be covered by e2e tests, but it would be nice if we could catch it already in a unit test. Could something like this work?: 
```go
t.Run("tiers", func(t *testing.T) {
	// given
	g, err := NewNSTemplateTierGenerator(s, Asset)
	require.NoError(t, err)

	// when
	tiers, err := g.NewNSTemplateTiers("my-cool-ns")
	require.NoError(t, err)
	require.NotEmpty(t, tiers)

	for _, actual := range tiers {
		tierName := actual.Name
		assert.Equal(t, "my-cool-ns", actual.Namespace)
		assert.Len(t, g.revisions[tierName], len(actual.Spec.Namespaces))
		for nsType, rev := range g.revisions[tierName] {
			found := false
			for _, ns := range actual.Spec.Namespaces {
				if ns.Type == nsType {
					found = true
					assert.Equal(t, rev, ns.Revision)
					asset, err := Asset(fmt.Sprintf("%s-%s.yaml", tierName, nsType))
					require.NoError(t, err)
					template, err := decodeTemplate(serializer.NewCodecFactory(s).UniversalDeserializer(), asset)
					require.NoError(t, err)
					assert.Equal(t, *template, ns.Template)
					break
				}
			}
			assert.True(t, found, "the namespace with the type wasn't found", "ns-type", nsType)
		}
	}
})
```

Reply:
ok, let me add something like that

Reply:
@MatousJobanek after giving it some thoughts, I've refactored the code and the tests in eca6ff9 so that now the `NSTemplateTierGenerator` and all its methods are package-protected, and all tests moved in the "whitebox" test file. In the next PR, I'll add a (unique) `CreateOrUpdateResources(...)` which will use this generator, but internally, so it won't be exposed to the outside world. Also, having all the tests in `pkg/templates/nstemplatetiers/nstemplatetier_generator_whitebox_test.go` makes it easier to maintain, and allows for accessing private fields such as `nstemplatetierGenerator.revisions` when asserting the results with the production data.

Please let me know what you think about it.

Reply:
I was thinking about such tests for the prod templates too but look.. This PR introduces NSTemplateTier generator which can work with any templates. And its functionality covered by the tests (using test templates). So we can be sure that the generator itself works as expected.

So, it's OK if we don't test the production templates themselves here. We should test them of course but it can/should be covered the next PRs when we will pass the prod templates to the generator (we don't yet). Basically it should be covered by `CreateOrUpdateResources` tests.